# Module 06 Lab - Regression and Classification Models

**Objective:** To understand the difference between regression and classification and to build your first linear models for both tasks.

**In this lab, you will write the code to train and evaluate the models.**

## Part 1: Regression vs. Classification

This is the most fundamental distinction between types of supervised learning problems.

*   **Regression:** The goal is to predict a **continuous numerical value.**
    *   *Examples:* Predicting the price of a house, the temperature tomorrow, or the stock price.

*   **Classification:** The goal is to predict a **discrete category or class label.**
    *   *Examples:* Predicting if an email is spam or not, a flower's species, or whether a customer will churn.

**In this lab, we will tackle one of each.**

## Part 2: Linear Regression

**Concept:** Linear Regression is used to predict a continuous value. It works by finding the best-fitting straight line through the data points. The model learns a "slope" (coefficient) for each feature and an "intercept".

*   **Problem:** We will predict the `Fare` of a Titanic passenger based on their `Age` and `Pclass`.

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Load and prepare the data
df = pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')

# For simplicity, we'll drop rows with missing age
df.dropna(subset=['Age'], inplace=True)

# Define features and target
features = ['Age', 'Pclass']
target_reg = 'Fare'

X_reg = df[features]
y_reg = df[target_reg]

# Split the data into training and test sets
X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(X_reg, y_reg, test_size=0.2, random_state=42)

### Task 1: Train and Evaluate a Linear Regression Model

**Your Task:**
1.  Create an instance of the `LinearRegression` model.
2.  Train the model using the training data (`X_train_reg`, `y_train_reg`).
3.  Make predictions on the test data.
4.  Evaluate the model using `mean_squared_error`.

In [2]:
# 1. Create the model instance
lr_model = LinearRegression()

# 2. Train (fit) the model on the training data
lr_model.fit(X_train_reg, y_train_reg)

# 3. Make predictions on the test data
y_pred_reg = lr_model.predict(X_test_reg)

# 4. Evaluate the model with mean squared error (average squared gap between predicted and actual fare)
mse = mean_squared_error(y_test_reg, y_pred_reg)
print(f"Mean Squared Error for Fare Prediction: {mse:.2f}")
print(f"The square root of this (RMSE) is {np.sqrt(mse):.2f}, so the model is off by about that many dollars on average.")

# For Knowledge Check Q2: print the learned coefficients (one slope per feature: Age, Pclass)
print("Coefficients (Age, Pclass):", lr_model.coef_)

Mean Squared Error for Fare Prediction: 3364.92
The square root of this (RMSE) is 58.01, so the model is off by about that many dollars on average.
Coefficients (Age, Pclass): [ -0.43695889 -38.29728935]


## Part 3: Logistic Regression

**Concept:** Despite its name, Logistic Regression is a **classification** algorithm. It calculates the probability that a given input belongs to a certain class. It is one of the most widely used and interpretable classification models.

*   **Problem:** We will predict whether a passenger `Survived` based on their `Age`, `Pclass`, and `Sex`.

In [3]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Prepare data for classification
# We need to encode the 'Sex' column into numbers
df['Sex'] = df['Sex'].map({'male': 0, 'female': 1})

# Define features and target
features_cls = ['Age', 'Pclass', 'Sex']
target_cls = 'Survived'

X_cls = df[features_cls]
y_cls = df[target_cls]

# Split the data
X_train_cls, X_test_cls, y_train_cls, y_test_cls = train_test_split(X_cls, y_cls, test_size=0.2, random_state=42)

### Task 2: Train and Evaluate a Logistic Regression Model

**Your Task:**
1.  Create an instance of the `LogisticRegression` model.
2.  Train the model using the classification training data.
3.  Make predictions on the test data.
4.  Evaluate the model using `accuracy_score`.

In [4]:
# 1. Create the model instance (max_iter raised so the model has enough steps to converge)
log_model = LogisticRegression(max_iter=1000)

# 2. Train (fit) the model on the classification training data
log_model.fit(X_train_cls, y_train_cls)

# 3. Make predictions on the test data
y_pred_cls = log_model.predict(X_test_cls)

# 4. Evaluate the model with accuracy (the share of passengers it classified correctly)
accuracy = accuracy_score(y_test_cls, y_pred_cls)
print(f"Accuracy for Survival Prediction: {accuracy:.2%}")

Accuracy for Survival Prediction: 74.83%


## 📝 Knowledge Check

**Instructions:** Answer the following questions in this markdown cell, in your own words.

1.  **In your own words, what is the key difference between a regression problem and a classification problem?**

*Your answer:* The difference comes down to what kind of value you are trying to predict. In **regression** the target is a continuous number that can fall anywhere on a scale, so the model can output things like 31.50 or 7.25. That is why predicting a passenger's `Fare` was a regression task — a fare is just a dollar amount. In **classification** the target is a discrete category from a fixed set of options, so the model picks a label rather than a number. Predicting `Survived` was classification because there are only two possible answers (survived = 1 or did not = 0). Short version: regression answers "how much?" and classification answers "which class?".

2.  **The `LinearRegression` model has an attribute called `.coef_`. After training, the code prints `lr_model.coef_`. What do these numbers represent?**

*Your answer:* They are the slopes (weights) the model learned for each feature, in the same order the features were listed — so `[-0.43695889, -38.29728935]` is the coefficient for `Age` first and `Pclass` second. Each one tells you how much the predicted `Fare` is expected to change when that feature goes up by 1, while the other feature stays the same. For `Age`, the value of about **-0.44** means every extra year of age lowers the predicted fare by roughly 44 cents. For `Pclass`, the value of about **-38.30** means moving up one class number (1st → 2nd → 3rd) lowers the predicted fare by about \$38.30. Both are negative, so both features have an inverse relationship with fare. The `Pclass` result makes sense because 1st class is the most expensive, so a *higher* class number actually means a *cheaper* ticket.

3.  **Why did we use `mean_squared_error` to evaluate the regression model but `accuracy_score` for the classification model? Why wouldn't accuracy be a good metric for the fare prediction task?**

*Your answer:* The metric has to match the type of output. `mean_squared_error` measures the average squared gap between the predicted number and the real number, which is exactly what you want when the answer is a continuous value like fare — it tells you how *close* you got. `accuracy_score` measures the fraction of predictions that landed in the correct category, which is the right idea when the answer is a class label like survived/did not. Accuracy would be a terrible metric for fare prediction because fare is continuous: the model almost never predicts the exact fare down to the cent, so an exact-match score would be near 0% and would tell us nothing useful — being off by 1 cent would count the same as being off by \$200. MSE (and its square root RMSE of about \$58 in my output) captures *how far off* the predictions are, which is what actually matters for a number.
